### PR TITLE
Feature: update new version of dav_subspace with higher performance

### PR DESCRIPTION
**Pull request by dyzheng**
### Reminder
- [ ] Have you linked an issue with this pull request?
- [ ] Have you added adequate unit tests and/or case tests for your pull request?
- [ ] Have you noticed possible changes of behavior below or in the linked issue?
- [ ] Have you explained the changes of codes in core modules of ESolver, HSolver, ElecState, Hamilt, Operator or Psi? (ignore if not applicable)

### Linked Issue
Fix #...

### Unit Tests and/or Case Tests for my changes
- A unit test is added for each new feature or bug fix.

### What's changed?
- Example: My changes might affect the performance of the application under certain conditions, and I have tested the impact on various scenarios...

### Any changes of core modules? (ignore if not applicable)
- Example: I have added a new virtual function in the esolver base class in order to ...

### CODE REVIEW

**Review by haozhihan — APPROVED** [review]

File: source/module_elecstate/potentials/pot_local.cpp [hunk 1]
```diff
@@ -29,6 +31,8 @@ void PotLocal::cal_fixed_v(double *vl_pseudo // store the local pseudopotential
         }
     }
 
+    PotLocal::vl_of_0 = vg[0].real();
```
Comment (mohanchen) [inline source/module_elecstate/potentials/pot_local.cpp]: what if vg array hasn't been initialized?

File: source/module_hsolver/hsolver_pw.cpp [hunk 2]
```diff
@@ -208,12 +209,45 @@ void HSolverPW<T, Device>::paw_func_after_kloop(psi::Psi<T, Device>& psi, elecst
...
 #endif
 
+template <typename T, typename Device>
+void HSolverPW<T, Device>::cal_ethr_band(const double& wk, const double* wg, const double& ethr, std::vector<double>& ethrs)
+{
+    if(wk > 0.0)
+    {
+        // Note: the idea of threshold for unoccupied bands (1e-5) comes from QE
+        // In ABACUS, We applied a smoothing process to this truncation to avoid abrupt changes in energy errors between different bands.
+        const double ethr_unocc = std::max(1e-5, ethr);
```
Comment (mohanchen) [inline source/module_hsolver/hsolver_pw.cpp]: there are three magic numbers, 1e-5, 1e-2, 1e-5 in these lines, could you use three variables to store these values in front of the code if(wk>0.0), and use these variables instead?

File: source/module_hsolver/hsolver_pw.cpp [hunk 3]
```diff
@@ -574,7 +617,7 @@ void HSolverPW<T, Device>::update_precondition(std::vector<Real>& h_diag, const
 
             if (this->method == "dav_subspace")
             {
-                h_diag[ig] = g2kin;
+                h_diag[ig] = g2kin + Real(elecstate::PotLocal::get_vl_of_0());
```
Comment (mohanchen) [inline source/module_hsolver/hsolver_pw.cpp]: would it be better to pass in the vl(0) value from the parameters of this function?

File: source/module_elecstate/potentials/pot_local.cpp [hunk 4]
```diff
@@ -29,6 +29,12 @@ void PotLocal::cal_fixed_v(double *vl_pseudo // store the local pseudopotential
         }
     }
 
+    /// save the V_local at G=0
+    if(this->rho_basis_->npw > 0)
+    {
+        *vl_of_0_ = vg[0].real();
+    }
```
Comment (mohanchen) [inline source/module_elecstate/potentials/pot_local.cpp:36]: is it possible to use such as  if (vg!=nullptr) instead of using npw?

Comment (dyzheng) [reply source/module_elecstate/potentials/pot_local.cpp:36]:
```
    std::complex<double> *vg = new std::complex<double>[this->rho_basis_->npw];
    ModuleBase::GlobalFunc::ZEROS(vg, this->rho_basis_->npw);
    for (int it = 0; it < this->ntype_; it++)
    {
        for (int ig = 0; ig < this->rho_basis_->npw; ig++)
        {
            vg[ig] += this->vloc_[0](it, this->rho_basis_->ig2igg[ig]) * this->sf_[0](it, ig);
        }
    }

//-------------------------------------------------
    /// save the V_local at G=0
    if(this->rho_basis_->npw > 0)
    {
        *vl_of_0_ = vg[0].real();
    }
//-------------------------------------------------

    const_cast<ModulePW::PW_Basis *>(this->rho_basis_)->recip2real(vg, vl_pseudo);
    delete[] vg;
```

Comment (dyzheng) [reply source/module_elecstate/potentials/pot_local.cpp:36]: @mohanchen  This is the context of this code.

**Review by mohanchen — APPROVED** [review]

**Pull request merged**